### PR TITLE
sanity check date prior to submission

### PR DESCRIPTION
**Pull request by changukshin**
- write one function to do sanity check (date_utils.check_date)
- add one option to config (max_submit_within_days)
- logics for version compatibility
    - unfamilliar with this implementation. may need to be modified.

### CODE REVIEW

File: tp_timesheet/date_utils.py [hunk 1]
```diff
@@ -48,3 +50,23 @@ def get_start_date(start_date_arg):
...
     return start_date
+
+
+def check_date(start_date):
+    """
+    check the start date submitting is bounded within `n` days.
+    parse `n` from config file. default : 7
+    """
+    start_date_str = start_date.strftime("%d/%m/%Y")
+    config = Config()
```
Comment (ThorpeJosh) [inline tp_timesheet/date_utils.py:61]: I don't think we should create another config object as `Config.__init__` performs read/writes to disk.
`Config.MAX_DAYS` should return what we are after as it is a class object

File: tp_timesheet/date_utils.py [hunk 2]
```diff
@@ -48,3 +50,23 @@ def get_start_date(start_date_arg):
...
+    today = datetime.today()
+    if int(config.MAX_DAYS) < abs(today - start_date).days:
+        user_confirm = input(
+            f"The start date you process is '{start_date_str}' "
+            f"and submitting beyond '{config.MAX_DAYS}' day is not recommended.\n"
```
Comment (ThorpeJosh) [inline tp_timesheet/date_utils.py:66]:
```suggestion
            f"The entered date '{start_date_str}' "
            f"is beyond the '{config.MAX_DAYS}' day window from today.\n"
```

File: tp_timesheet/__main__.py [hunk 3]
```diff
@@ -82,6 +82,7 @@ def run():
     cal = Singapore()
 
     start_date = get_start_date(args.start)
+    check_date(start_date)
```
Comment (ThorpeJosh) [inline tp_timesheet/__main__.py:85]: maybe to make it more obvious what this function does and that it doesn't return anything, we can call it `assert_start_date`

**Review by ThorpeJosh — CHANGES_REQUESTED** [review]

File: tp_timesheet/date_utils.py [hunk 4]
```diff
@@ -48,3 +50,23 @@ def get_start_date(start_date_arg):
...
+            "You can modify the threshold and/or disable this confirmation option from '{config.CONFIG_PATH}'.\n"
```
[the remark this answers is not in the document]
Comment (changukshin) [reply tp_timesheet/date_utils.py]: Addressed on f5d3ef4.

File: tp_timesheet/date_utils.py (hunk 2 above)
Comment (changukshin) [reply tp_timesheet/date_utils.py]: Addressed on f5d3ef4.

File: tp_timesheet/date_utils.py [hunk 5]
```diff
@@ -48,3 +50,23 @@ def get_start_date(start_date_arg):
...
+            f"Do you want to submit the timesheet of the date?[y/N]: "
```
[the remark this answers is not in the document]
Comment (changukshin) [reply tp_timesheet/date_utils.py]: Addressed on f5d3ef4.

File: tp_timesheet/__main__.py (hunk 3 above)
Comment (changukshin) [reply tp_timesheet/__main__.py]: Addressed on f5d3ef4.

**Review by ThorpeJosh — CHANGES_REQUESTED** [review]
Looks good. Just needs to be rebased as `main` has changed causing merge conflicts. (this is why Jenkins cannot run the `pr-merge` check)

File: tp_timesheet/config.py [hunk 6]
```diff
@@ -115,4 +119,13 @@ def _read_write_config(cls):
             logger.debug("Reading config file at: %s", cls.CONFIG_PATH)
         input_config = configparser.ConfigParser()
         input_config.read(cls.CONFIG_PATH)
+
+        # Version compatibility (#20)
+        for config_key in ["check_max_days", "max_submit_within_days"]:
```
Comment (ThorpeJosh) [inline tp_timesheet/config.py:124]:
```suggestion
        for config_key in cls.DEFAULT:
```
Shouldn't have these values hard coded twice. Can iterate over the dictionary keys like this.

File: tp_timesheet/date_utils.py [hunk 7]
```diff
@@ -48,3 +50,24 @@ def get_start_date(start_date_arg):
             else:
                 start_date = cand2
     return start_date
+
+
+def assert_start_date(start_date):
```
Comment (ThorpeJosh) [inline tp_timesheet/date_utils.py:55]: Can you please add a unit test for this function?

**Review by ThorpeJosh — CHANGES_REQUESTED** [review]
Thanks for the changes @changukshin.

I'm being a bit picky now but would be great to have unit tests written for all new code now that we have a pipeline setup.

File: tp_timesheet/config.py [hunk 8]
```diff
@@ -115,4 +119,13 @@ def _read_write_config(cls):
             logger.debug("Reading config file at: %s", cls.CONFIG_PATH)
         input_config = configparser.ConfigParser()
         input_config.read(cls.CONFIG_PATH)
+
+        # Version compatibility (#20)
```
Comment (ThorpeJosh) [inline tp_timesheet/config.py:123]: Can we please write a unit to check the following cases:

case 1:
* Given no `tp.conf` exists
* When a `config.Config()` object is created
* Then `tp.conf` is created and populated with the default values

case 2:
* Given `tp.conf` exists
* When a `config.Config()` object is created
* Then `tp.conf` is updated with missing keys and their default values

**Review by ThorpeJosh — CHANGES_REQUESTED** [review]
Thanks for the changes @changukshin.

I'm being a bit picky now but would be great to have unit tests written for all new code now that we have a pipeline setup.

File: tp_timesheet/date_utils.py (hunk 7 above)
Comment (changukshin) [reply tp_timesheet/date_utils.py:55]: This function have to get users input to confirm.
Do you have any idea to pass imitated confirmation to the function when you test any function which get the keyboard input of user?

Comment (changukshin) [reply tp_timesheet/date_utils.py:55]: I found a way to manipulate. On progress.

File: tp_timesheet/config.py (hunk 8 above)
Comment (changukshin) [reply tp_timesheet/config.py:123]: On it!

**Comment by ThorpeJosh** [comment]
@changukshin You happy for me to rebase this?

**Comment by changukshin** [comment]
@ThorpeJosh Sure. Thank you. I have something to do so I have no slot to follow this. Sorry for this delay.

**Comment by ThorpeJosh** [comment]
> @ThorpeJosh Sure. Thank you. I have something to do so I have no slot to follow this. Sorry for this delay.

No stress. Raised a new PR #57 with this work

**Pull request closed without merging**